packages/graph/hash_graph/lib/graph/src/shared/subgraph/mod.rs

```diff
@@ -29,4 +38,60 @@ impl Subgraph {
             depths,
         }
     }
+
+    fn entry<R: Record>(
+        &mut self,
+        edition_id: &impl SubgraphIndex<R>,
+    ) -> RawEntryMut<R::EditionId, R, RandomState> {
+        edition_id.subgraph_entry(self)
+    }
+
+    pub fn insert<R: Record>(&mut self, record: R) -> Option<R> {
+        match self.entry(record.edition_id()) {
+            RawEntryMut::Occupied(mut entry) => Some(entry.insert(record)),
+            RawEntryMut::Vacant(entry) => {
+                entry.insert(record.edition_id().clone(), record);
+                None
+            }
+        }
+    }
+
+    /// Looks up a single [`Record`] in the subgraph or reads it from the [`Store`] and inserts it
+    /// if it is not yet in the subgraph.
+    ///
+    /// # Errors
+    ///
+    /// - Returns an error if the [`Record`] could not be read from the [`Store`].
+    ///
+    /// [`Store`]: crate::store::Store
+    pub async fn get_or_read<'r, R: Record + Sync + 'r>(
+        &'r mut self,
+        store: &impl Read<R>,
+        edition_id: &R::EditionId,
+    ) -> Result<&'r R, QueryError> {
+        Ok(match self.entry(edition_id) {
+            RawEntryMut::Occupied(entry) => entry.into_mut(),
+            RawEntryMut::Vacant(entry) => {
+                entry
+                    .insert(
+                        edition_id.clone(),
+                        store
+                            .read_one(&R::create_filter_for_edition_id(edition_id))
+                            .await?,
+                    )
+                    .1
+            }
+        })
+    }
+}
+
+/// Used for index operations on a mutable [`Subgraph`].
+///
+/// Depending on `R`, the index operation will be performed on the vertices of the subgraph.
```

Comment:
```suggestion
/// Depending on `R`, the index operation will be performed on the respective vertices collection of the subgraph.
```

Maybe something like that? I think the main thing we're trying to convey here is that this will go to `vertices.property_type` or `vertices.entity`, etc.

Also can this be used for Edges, or not? If it can't then maybe we should call this `SubgraphVertexIndex`

Reply:
We probably can add a new function to access the edges as well but currently, we don't require that. This would require a few adjustments to the `Record` trait (adding an associated type on the edge kind) and the `Subgraph` structure (similar to `Vertices` to directly allow accessing the different ontology types), but I don't see a reason to not add the function as soon as we need generic access.
Maybe this?
```rust
/// Used for index operations on a mutable [`Subgraph`].
///
/// Depending on `R`, the index operation will be performed on the respective collection of the
/// subgraph.
pub trait SubgraphIndex<R: Record>: Clone + Eq + Hash + Into<GraphElementEditionId> {
    /// Returns a mutable reference to the [`Record`] vertex in the subgraph.
    fn subgraph_vertex_entry<'a>(
        &self,
        subgraph: &'a mut Subgraph,
    ) -> RawEntryMut<'a, R::EditionId, R, RandomState>;
}
```

Reply:
Looks good to me

Reply:
Alright, changed it in 95d36c80bd790225412aaca6705703faf34f4c53